Compilation on Jetson Orin

### The bug

I tried to compile on a Jetson Orin. I changed the path of the compiler to 

I'm getting the following error:

```log
/home/pantelis/Documents/GPUtils/test/testTensor.cu(676): error: identifier "requires" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(676): error: namespace "std" has no member "floating_point"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(676): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(683): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(684): error: identifier "svd" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(685): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(685): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(686): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(686): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(689): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(689): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(689): warning #12-D: parsing restarts here after previous syntax error

/home/pantelis/Documents/GPUtils/test/testTensor.cu(690): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(693): error: identifier "singularValuesComputation" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(693): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(694): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(703): error: identifier "requires" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(703): error: namespace "std" has no member "floating_point"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(703): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(710): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(711): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(712): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(713): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(713): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(714): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(714): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(715): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(716): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(717): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(717): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(718): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(718): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(721): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(721): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(722): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(722): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(723): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(726): error: identifier "singularValuesMemory" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(726): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(727): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(735): error: identifier "requires" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(735): error: namespace "std" has no member "floating_point"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(735): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(750): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): error: identifier "T" is undefined

/usr/include/c++/9/bits/ptr_traits.h(119): error: static assertion failed with "pointer type defines element_type or is like SomePointer<T, Args>"
          detected during:
            instantiation of class "std::pointer_traits<_Ptr> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" 
/usr/include/c++/9/bits/alloc_traits.h(102): here
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__c_pointer, _Tp=const <error-type>, <unnamed>=void]" 
/usr/include/c++/9/bits/alloc_traits.h(135): here
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(84): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/ptr_traits.h(113): error: class "std::pointer_traits<_Ptr>::__rebind<std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>, const <error-type>, void> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" has no member "type"
          detected during:
            instantiation of type "std::pointer_traits<_Ptr>::rebind<const <error-type>> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" 
/usr/include/c++/9/bits/alloc_traits.h(102): here
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__c_pointer, _Tp=const <error-type>, <unnamed>=void]" 
/usr/include/c++/9/bits/alloc_traits.h(135): here
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(84): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/ptr_traits.h(113): error: class "std::pointer_traits<_Ptr>::__rebind<std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>, void, void> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" has no member "type"
          detected during:
            instantiation of type "std::pointer_traits<_Ptr>::rebind<void> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" 
/usr/include/c++/9/bits/alloc_traits.h(102): here
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__v_pointer, _Tp=void, <unnamed>=void]" 
/usr/include/c++/9/bits/alloc_traits.h(143): here
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(84): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/ptr_traits.h(113): error: class "std::pointer_traits<_Ptr>::__rebind<std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>, const void, void> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" has no member "type"
          detected during:
            instantiation of type "std::pointer_traits<_Ptr>::rebind<const void> [with _Ptr=std::__detected_or_t<<error-type> *, std::__allocator_traits_base::__pointer, std::allocator<<error-type>>>]" 
/usr/include/c++/9/bits/alloc_traits.h(102): here
            instantiation of class "std::allocator_traits<_Alloc>::_Ptr<_Func, _Tp, <unnamed>> [with _Alloc=std::allocator<<error-type>>, _Func=std::__allocator_traits_base::__cv_pointer, _Tp=const void, <unnamed>=void]" 
/usr/include/c++/9/bits/alloc_traits.h(151): here
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(84): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(76): error: class "std::__allocator_traits_base::__rebind<std::allocator<<error-type>>, <error-type>, void>" has no member "type"
          detected during:
            instantiation of type "std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>" 
(206): here
            instantiation of type "std::allocator_traits<_Alloc>::rebind_alloc<<error-type>> [with _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(120): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>>::rebind<_Tp> [with _Alloc=std::allocator<<error-type>>, <unnamed>=<error-type>, _Tp=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(84): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/ext/alloc_traits.h(49): error: a class or namespace qualified name is required
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>> 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/ext/alloc_traits.h(49): error: expected a ","
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>> 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(88): error: a class or namespace qualified name is required
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(88): error: global-scope qualifier (leading "::") is not allowed
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(88): error: expected a ";"
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(95): error: identifier "value_type" is undefined
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/alloc_traits.h(135): error: expected a ">"
          detected during:
            instantiation of class "std::allocator_traits<_Alloc> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>]" 
/usr/include/c++/9/ext/alloc_traits.h(52): here
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/ext/alloc_traits.h(58): error: class "std::allocator_traits<std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>>" has no member "value_type"
          detected during:
            instantiation of class "__gnu_cxx::__alloc_traits<_Alloc, <unnamed>> [with _Alloc=std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>, <unnamed>=<error-type>]" 
/usr/include/c++/9/bits/stl_vector.h(85): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/usr/include/c++/9/bits/stl_vector.h(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/stl_vector.h(126): error: not a class or struct name
          detected during:
            instantiation of class "std::_Vector_base<_Tp, _Alloc>::_Vector_impl [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
(337): here
            instantiation of class "std::_Vector_base<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
(386): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/stl_vector.h(398): error: static assertion failed with "std::vector must have a non-const, non-volatile value_type"
          detected during instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/bits/stl_vector.h(401): error: static assertion failed with "std::vector must have the same value_type as its allocator"
          detected during instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/ext/alloc_traits.h(49): error: a class or namespace qualified name is required
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>> 
/usr/include/c++/9/bits/stl_vector.h(408): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/usr/include/c++/9/ext/alloc_traits.h(49): error: expected a ","
          detected during:
            processing of template argument list for "__gnu_cxx::__alloc_traits" based on template argument <std::__alloc_rebind<std::allocator<<error-type>>, <error-type>>> 
/usr/include/c++/9/bits/stl_vector.h(408): here
            instantiation of class "std::vector<_Tp, _Alloc> [with _Tp=<error-type>, _Alloc=std::allocator<<error-type>>]" 
/home/pantelis/Documents/GPUtils/test/testTensor.cu(751): here

/home/pantelis/Documents/GPUtils/test/testTensor.cu(752): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(753): error: this declaration has no storage class or type specifier

/home/pantelis/Documents/GPUtils/test/testTensor.cu(753): error: variable "S" has already been defined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(753): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(754): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(754): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(755): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(766): error: identifier "T" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(767): error: this declaration has no storage class or type specifier

/home/pantelis/Documents/GPUtils/test/testTensor.cu(767): error: variable "U" has already been defined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(767): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(768): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(768): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(770): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(773): error: identifier "singularValuesMultipleMatrices" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(773): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(774): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(783): error: identifier "requires" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(783): error: namespace "std" has no member "floating_point"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(783): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(793): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(794): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(794): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(795): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(795): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(796): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(799): error: identifier "singularValuesRankMultipleMatrices" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(799): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(800): error: type name is not allowed

/home/pantelis/Documents/GPUtils/test/testTensor.cu(819): error: identifier "requires" is undefined

/home/pantelis/Documents/GPUtils/test/testTensor.cu(819): error: namespace "std" has no member "floating_point"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(819): error: expected a ";"

/home/pantelis/Documents/GPUtils/test/testTensor.cu(827): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(828): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(828): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(829): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(829): error: expected a declaration

/home/pantelis/Documents/GPUtils/test/testTensor.cu(830): error: expected a declaration

Error limit reached.
100 errors detected in the compilation of "/home/pantelis/Documents/GPUtils/test/testTensor.cu".
Compilation terminated.
make[2]: *** [test/CMakeFiles/device_test.dir/build.make:77: test/CMakeFiles/device_test.dir/testTensor.cu.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:180: test/CMakeFiles/device_test.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I believe we've left some C++20 features here and there.


### Platform information 

 - OS: Ubuntu 23
 - NVCC version 11.4


For completeness, this is the modified `CMakeLists.txt`:

```
# ====================================================================
# GPUtils
# ====================================================================
cmake_minimum_required(VERSION 3.20 FATAL_ERROR)
project(GPUtils
    DESCRIPTION "Easy use of vectors and matrices on GPGPU devices."
    HOMEPAGE_URL "https://github.com/GPUEngineering/GPUtils"
    LANGUAGES CXX
)
# ----
set(CMAKE_CUDA_SEPARABLE_COMPILATION ON)  # required for calling cuda kernels from cuda kernels
set(CMAKE_CUDA_COMPILER "/usr/local/cuda-11.4/bin/nvcc")
set(CMAKE_CUDA_ARCHITECTURES 86)
set(CMAKE_CUDA_STANDARD 20)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CUDA_FLAGS "-std=c++17")
set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS}; -std=c++17)
enable_language(CUDA)
# ----
add_library(device_compiler_flags INTERFACE)
target_compile_features(device_compiler_flags INTERFACE cxx_std_17)
set(CMAKE_CXX_EXTENSIONS OFF)
# ----
add_library(developer_flags INTERFACE)
set(cxx_flags -Wall)
set(cuda_flags -arch=sm_60 -std=c++17 -Xcompiler=-Wall -Xcudafe=--display_error_number -g)
target_compile_options(developer_flags
    INTERFACE
    # flags for CXX builds
    $<$<COMPILE_LANGUAGE:CXX>:${cxx_flags}>
    # flags for CUDA builds
    $<$<COMPILE_LANGUAGE:CUDA>:${cuda_flags}>
)
target_link_libraries(device_compiler_flags INTERFACE $<BUILD_INTERFACE:developer_flags>)
# ----


# ====================================================================
# comment out for release
# ====================================================================
add_executable(main)
target_sources(main
    PRIVATE
    main.cu
)
target_link_libraries(main
    PRIVATE
    device_compiler_flags
    cublas
    cusolver
    cudadevrt
)
target_include_directories(main
    PRIVATE
    "${PROJECT_BINARY_DIR}"
    "${PROJECT_SOURCE_DIR}/include"
)
# ----
add_subdirectory(test)
# ----
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -7,7 +7,10 @@ name: Continuous integration
 
 jobs:
   ci:
-    runs-on: [self-hosted, orin]
+    runs-on: ${{ matrix.runner }}
+    strategy:
+      matrix:
+        runner: [alphaville, pop-os]
     steps:
       - name: checkout code
         uses: actions/checkout@v4

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -7,7 +7,7 @@ name: Continuous integration
 
 jobs:
   ci:
-    runs-on: self-hosted
+    runs-on: [self-hosted, orin]
     steps:
       - name: checkout code
         uses: actions/checkout@v4

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,7 +10,7 @@ jobs:
     runs-on: ${{ matrix.runner }}
     strategy:
       matrix:
-        runner: [alphaville, pop-os]
+        runner: [orin, a40]
     steps:
       - name: checkout code
         uses: actions/checkout@v4

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,36 +2,52 @@
 # GPUtils
 # ====================================================================
 cmake_minimum_required(VERSION 3.20 FATAL_ERROR)
+cmake_policy(SET CMP0135 NEW)
 project(GPUtils
-    DESCRIPTION "Easy use of vectors and matrices on GPGPU devices."
-    HOMEPAGE_URL "https://github.com/GPUEngineering/GPUtils"
-    LANGUAGES CXX
-)
+        DESCRIPTION "Easy use of vectors and matrices on GPGPU devices."
+        HOMEPAGE_URL "https://github.com/GPUEngineering/GPUtils"
+        LANGUAGES CXX
+        )
+
+
+option(CPPVERSION "C++ version" 20) # A40: 20, Orin: 17
+option(SM_ARCH "SM architecture" 86) # A40: 86, Orin: 87
+
+set (cppversion ${CPPVERSION})
+set (cppstd "c++${CPPVERSION}")
+set (cxxstd cxx_std_${CPPVERSION})
+
 # ----
+
 set(CMAKE_CUDA_SEPARABLE_COMPILATION ON)  # required for calling cuda kernels from cuda kernels
-set(CMAKE_CUDA_COMPILER "/usr/local/cuda-12.3/bin/nvcc")
-set(CMAKE_CUDA_ARCHITECTURES 86)
-set(CMAKE_CUDA_STANDARD 20)
-set(CMAKE_CXX_STANDARD 20)
-set(CMAKE_CUDA_FLAGS "-std=c++20")
-set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS}; -std=c++20)
+set(CMAKE_CUDA_COMPILER "/usr/local/cuda-11.4/bin/nvcc")
+set(CMAKE_CUDA_ARCHITECTURES ${SM_ARCH})
+set(CMAKE_CUDA_STANDARD ${cppversion})
+set(CMAKE_CXX_STANDARD ${cppversion})
+set(CMAKE_CUDA_FLAGS "-std=${cppstd}")
+set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS}; -std=${cppstd})
 enable_language(CUDA)
+
 # ----
+
 add_library(device_compiler_flags INTERFACE)
-target_compile_features(device_compiler_flags INTERFACE cxx_std_20)
+target_compile_features(device_compiler_flags INTERFACE ${cxxstd})
 set(CMAKE_CXX_EXTENSIONS OFF)
+
 # ----
+
 add_library(developer_flags INTERFACE)
 set(cxx_flags -Wall)
-set(cuda_flags -arch=sm_60 -std=c++20 -Xcompiler=-Wall -Xcudafe=--display_error_number -g)
+set(cuda_flags -arch=sm_${SM_ARCH} -std=${cppstd} -Xcompiler=-Wall -Xcudafe=--display_error_number -g)
 target_compile_options(developer_flags
-    INTERFACE
-    # flags for CXX builds
-    $<$<COMPILE_LANGUAGE:CXX>:${cxx_flags}>
-    # flags for CUDA builds
-    $<$<COMPILE_LANGUAGE:CUDA>:${cuda_flags}>
-)
+        INTERFACE
+        # flags for CXX builds
+        $<$<COMPILE_LANGUAGE:CXX>:${cxx_flags}>
+        # flags for CUDA builds
+        $<$<COMPILE_LANGUAGE:CUDA>:${cuda_flags}>
+        )
 target_link_libraries(device_compiler_flags INTERFACE $<BUILD_INTERFACE:developer_flags>)
+
 # ----
 
 
@@ -40,21 +56,21 @@ target_link_libraries(device_compiler_flags INTERFACE $<BUILD_INTERFACE:develope
 # ====================================================================
 add_executable(main)
 target_sources(main
-    PRIVATE
-    main.cu
-)
+        PRIVATE
+        main.cu
+        )
 target_link_libraries(main
-    PRIVATE
-    device_compiler_flags
-    cublas
-    cusolver
-    cudadevrt
-)
+        PRIVATE
+        device_compiler_flags
+        cublas
+        cusolver
+        cudadevrt
+        )
 target_include_directories(main
-    PRIVATE
-    "${PROJECT_BINARY_DIR}"
-    "${PROJECT_SOURCE_DIR}/include"
-)
+        PRIVATE
+        "${PROJECT_BINARY_DIR}"
+        "${PROJECT_SOURCE_DIR}/include"
+        )
 # ----
 add_subdirectory(test)
 # ----

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,43 +2,42 @@
 # GPUtils
 # ====================================================================
 cmake_minimum_required(VERSION 3.20 FATAL_ERROR)
-cmake_policy(SET CMP0135 NEW)
+
+if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.29")
+    cmake_policy(SET CMP0135 NEW)
+endif()
+
+# Set C++ version and SM architecture
+if (NOT DEFINED CPPVERSION)
+    set(CPPVERSION 20) # A40: 20, Orin: 17
+endif()
+if (NOT DEFINED SM_ARCH)
+    set(SM_ARCH 86)# A40: 86, Orin: 87
+endif()
+
+
 project(GPUtils
         DESCRIPTION "Easy use of vectors and matrices on GPGPU devices."
         HOMEPAGE_URL "https://github.com/GPUEngineering/GPUtils"
         LANGUAGES CXX
         )
-
-
-option(CPPVERSION "C++ version" 20) # A40: 20, Orin: 17
-option(SM_ARCH "SM architecture" 86) # A40: 86, Orin: 87
-
-set (cppversion ${CPPVERSION})
-set (cppstd "c++${CPPVERSION}")
-set (cxxstd cxx_std_${CPPVERSION})
-
 # ----
-
 set(CMAKE_CUDA_SEPARABLE_COMPILATION ON)  # required for calling cuda kernels from cuda kernels
-set(CMAKE_CUDA_COMPILER "/usr/local/cuda-11.4/bin/nvcc")
+set(CMAKE_CUDA_COMPILER "/usr/local/cuda/bin/nvcc")
 set(CMAKE_CUDA_ARCHITECTURES ${SM_ARCH})
-set(CMAKE_CUDA_STANDARD ${cppversion})
-set(CMAKE_CXX_STANDARD ${cppversion})
-set(CMAKE_CUDA_FLAGS "-std=${cppstd}")
-set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS}; -std=${cppstd})
+set(CMAKE_CUDA_STANDARD ${CPPVERSION})
+set(CMAKE_CXX_STANDARD ${CPPVERSION})
+set(CMAKE_CUDA_FLAGS "-std=c++${CPPVERSION}")
+set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS}; "-std=c++${CPPVERSION}")
 enable_language(CUDA)
-
 # ----
-
 add_library(device_compiler_flags INTERFACE)
-target_compile_features(device_compiler_flags INTERFACE ${cxxstd})
+target_compile_features(device_compiler_flags INTERFACE cxx_std_${CPPVERSION})
 set(CMAKE_CXX_EXTENSIONS OFF)
-
 # ----
-
 add_library(developer_flags INTERFACE)
 set(cxx_flags -Wall)
-set(cuda_flags -arch=sm_${SM_ARCH} -std=${cppstd} -Xcompiler=-Wall -Xcudafe=--display_error_number -g)
+set(cuda_flags -arch=sm_${SM_ARCH} -std=c++${CPPVERSION} -Xcompiler=-Wall -Xcudafe=--display_error_number -g)
 target_compile_options(developer_flags
         INTERFACE
         # flags for CXX builds
@@ -47,7 +46,6 @@ target_compile_options(developer_flags
         $<$<COMPILE_LANGUAGE:CUDA>:${cuda_flags}>
         )
 target_link_libraries(device_compiler_flags INTERFACE $<BUILD_INTERFACE:developer_flags>)
-
 # ----
 
 

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -4,6 +4,8 @@ set -euxo pipefail
 
 tests() {
     # Where are we? (A40 or Orin?)
+    cpp_version=17 # default
+    sm_arch=86 # default
     hwInfoOrin=`lshw | grep Orin` ||
     if [ ! -z "$(hwInfoOrin)" ]; then
       echo "Running on Orin";

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -3,7 +3,7 @@ set -euxo pipefail
 
 tests() {
     # Where are we?
-    hwInfoOrin = `lshw | grep Orin`
+    hwInfoOrin=`lshw | grep Orin`
     if [ ! -z "$hwInfoOrin" ]; then
       echo "Running on Orin";
       sm_arch=87

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -2,12 +2,24 @@
 set -euxo pipefail
 
 tests() {
+    # Where are we?
+    hwInfoOrin = `lshw | grep Orin`
+    if [ ! -z "$hwInfoOrin" ]; then
+      echo "Running on Orin";
+      sm_arch=87
+      cpp_version=17
+    else
+      echo "Not running on Orin";
+      sm_arch=86
+      cpp_version=20
+    fi
+
     # ------------------------------------
     # Run tensor gtests
     # ------------------------------------
 
     # -- create build files
-    cmake -S . -B ./build -Wno-dev
+    cmake -DCPPVERSION=${cpp_version} -DSM_ARCH=${sm_arch} -S . -B ./build -Wno-dev
 
     # -- build files in build folder
     cmake --build ./build
@@ -17,7 +29,7 @@ tests() {
 
     # -- run compute sanitizer
     cd ./build/test
-    mem=$(/usr/local/cuda-12.3/bin/compute-sanitizer --tool memcheck --leak-check=full ./device_test)
+    mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./device_test)
     grep "0 errors" <<< "$mem"
     cd ../..
 
@@ -27,7 +39,7 @@ tests() {
 
     # -- create build files
     cd example
-    cmake -S . -B ./build -Wno-dev
+    cmake  -DCPPVERSION=${cpp_version} -DSM_ARCH=${sm_arch} -S . -B ./build -Wno-dev
 
     # -- build files in build folder
     cmake --build ./build
@@ -37,7 +49,7 @@ tests() {
 
     # -- run compute sanitizer
     cd ./build
-    mem=$(/usr/local/cuda-12.3/bin/compute-sanitizer --tool memcheck --leak-check=full ./example_main)
+    mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./example_main)
     grep "0 errors" <<< "$mem"
 }
 

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -7,7 +7,7 @@ tests() {
     cpp_version=17 # default
     sm_arch=86 # default
     hwInfoOrin=`lshw | grep Orin` ||
-    if [ ! -z "$(hwInfoOrin)" ]; then
+    if [ ! -z "${hwInfoOrin}" ]; then
       echo "Running on Orin";
       sm_arch=87
       cpp_version=17
@@ -30,7 +30,7 @@ tests() {
     # -- run tests
     ctest --test-dir ./build/test --output-on-failure
 
-    if [ -z "$(hwInfoOrin)" ]; then
+    if [ -z "${hwInfoOrin}" ]; then
 
       # -- run compute sanitizer
       cd ./build/test

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -1,6 +1,7 @@
 #!/bin/bash
 set -euxo pipefail
 
+
 tests() {
     # Where are we? (A40 or Orin?)
     hwInfoOrin=`lshw | grep Orin` ||

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -2,9 +2,9 @@
 set -euxo pipefail
 
 tests() {
-    # Where are we?
-    hwInfoOrin=`lshw | grep Orin`
-    if [ ! -z "$hwInfoOrin" ]; then
+    # Where are we? (A40 or Orin?)
+    hwInfoOrin=`lshw | grep Orin` ||
+    if [ ! -z "$(hwInfoOrin)" ]; then
       echo "Running on Orin";
       sm_arch=87
       cpp_version=17

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -30,6 +30,10 @@ tests() {
     # -- run tests
     ctest --test-dir ./build/test --output-on-failure
 
+    if [ ! -z "$(hwInfoOrin)" ]; then
+      return;
+    fi
+
     # -- run compute sanitizer
     cd ./build/test
     mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./device_test)

--- a/ci/script.sh
+++ b/ci/script.sh
@@ -30,34 +30,33 @@ tests() {
     # -- run tests
     ctest --test-dir ./build/test --output-on-failure
 
-    if [ ! -z "$(hwInfoOrin)" ]; then
-      return;
+    if [ -z "$(hwInfoOrin)" ]; then
+
+      # -- run compute sanitizer
+      cd ./build/test
+      mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./device_test)
+      grep "0 errors" <<< "$mem"
+      cd ../..
+
+      # ------------------------------------
+      # Run example executable
+      # ------------------------------------
+
+      # -- create build files
+      cd example
+      cmake  -DCPPVERSION=${cpp_version} -DSM_ARCH=${sm_arch} -S . -B ./build -Wno-dev
+
+      # -- build files in build folder
+      cmake --build ./build
+
+      # -- run main.cu
+      ./build/example_main
+
+      # -- run compute sanitizer
+      cd ./build
+      mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./example_main)
+      grep "0 errors" <<< "$mem"
     fi
-
-    # -- run compute sanitizer
-    cd ./build/test
-    mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./device_test)
-    grep "0 errors" <<< "$mem"
-    cd ../..
-
-    # ------------------------------------
-    # Run example executable
-    # ------------------------------------
-
-    # -- create build files
-    cd example
-    cmake  -DCPPVERSION=${cpp_version} -DSM_ARCH=${sm_arch} -S . -B ./build -Wno-dev
-
-    # -- build files in build folder
-    cmake --build ./build
-
-    # -- run main.cu
-    ./build/example_main
-
-    # -- run compute sanitizer
-    cd ./build
-    mem=$(/usr/local/cuda/bin/compute-sanitizer --tool memcheck --leak-check=full ./example_main)
-    grep "0 errors" <<< "$mem"
 }
 
 

--- a/include/tensor.cuh
+++ b/include/tensor.cuh
@@ -1358,7 +1358,7 @@ inline void CholeskyBatchFactoriser<double>::solve(DTensor<double> &b) {
     if (!m_factorisationDone) throw std::logic_error("[CholeskyBatchSolve] no factor to solve with");
     if (m_numRows != b.numRows() || m_numMats != b.numMats()) {
         throw std::invalid_argument("[CholeskyBatchSolve] A and b incompatible");
-
+    }
     if (b.numCols() != 1) throw std::invalid_argument("[CholeskyBatchSolve] only supports `b` with one column");
     DTensor<double *> ptrA = m_matrix->pointersToMatrices();
     DTensor<double *> ptrB = b.pointersToMatrices();

--- a/include/tensor.cuh
+++ b/include/tensor.cuh
@@ -1326,7 +1326,7 @@ public:
 };
 
 template<>
-void CholeskyBatchFactoriser<double>::factorise() {
+inline void CholeskyBatchFactoriser<double>::factorise() {
     if (m_factorisationDone) return;
     DTensor<double *> ptrA = m_matrix->pointersToMatrices();
     gpuErrChk(cusolverDnDpotrfBatched(Session::getInstance().cuSolverHandle(),
@@ -1340,7 +1340,7 @@ void CholeskyBatchFactoriser<double>::factorise() {
 }
 
 template<>
-void CholeskyBatchFactoriser<float>::factorise() {
+inline void CholeskyBatchFactoriser<float>::factorise() {
     if (m_factorisationDone) return;
     DTensor<float *> ptrA = m_matrix->pointersToMatrices();
     gpuErrChk(cusolverDnSpotrfBatched(Session::getInstance().cuSolverHandle(),
@@ -1354,8 +1354,11 @@ void CholeskyBatchFactoriser<float>::factorise() {
 }
 
 template<>
-void CholeskyBatchFactoriser<double>::solve(DTensor<double> &b) {
+inline void CholeskyBatchFactoriser<double>::solve(DTensor<double> &b) {
     if (!m_factorisationDone) throw std::logic_error("[CholeskyBatchSolve] no factor to solve with");
+    if (m_numRows != b.numRows() || m_numMats != b.numMats()) {
+        throw std::invalid_argument("[CholeskyBatchSolve] A and b incompatible");
+
     if (b.numCols() != 1) throw std::invalid_argument("[CholeskyBatchSolve] only supports `b` with one column");
     DTensor<double *> ptrA = m_matrix->pointersToMatrices();
     DTensor<double *> ptrB = b.pointersToMatrices();
@@ -1372,8 +1375,11 @@ void CholeskyBatchFactoriser<double>::solve(DTensor<double> &b) {
 }
 
 template<>
-void CholeskyBatchFactoriser<float>::solve(DTensor<float> &b) {
+inline void CholeskyBatchFactoriser<float>::solve(DTensor<float> &b) {
     if (!m_factorisationDone) throw std::logic_error("[CholeskyBatchSolve] no factor to solve with");
+    if (m_numRows != b.numRows() || m_numMats != b.numMats()) {
+        throw std::invalid_argument("[CholeskyBatchSolve] A and b incompatible");
+    }
     if (b.numCols() != 1) throw std::invalid_argument("[CholeskyBatchSolve] only supports `b` with one column");
     DTensor<float *> ptrA = m_matrix->pointersToMatrices();
     DTensor<float *> ptrB = b.pointersToMatrices();

--- a/test/testTensor.cu
+++ b/test/testTensor.cu
@@ -24,7 +24,7 @@ protected:
  * Zero Tensor (Constructor)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorConstructionZero() {
     DTensor<T> zero(2, 3, 4, true);
     EXPECT_EQ(2, zero.numRows());
@@ -46,7 +46,7 @@ TEST_F(TensorTest, tensorConstructionZero) {
  * Row- and column-major data
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorConstructionStorageMode() {
     size_t rows = 3;
     size_t cols = 2;
@@ -98,7 +98,7 @@ TEST_F(TensorTest, tensorConstructionStorageMode) {
  * Move constructor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorMoveConstructor() {
     DTensor<T> zero(2, 3, 4, true);
     DTensor<T> x(std::move(zero));
@@ -118,7 +118,7 @@ TEST_F(TensorTest, tensorMoveConstructor) {
  * Constructor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorConstructionFromVector() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -138,7 +138,7 @@ TEST_F(TensorTest, tensorConstructionFromVector) {
  * Tensor: Copy constructor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorCopyConstructor() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -164,7 +164,7 @@ TEST_F(TensorTest, tensorCopyConstructor) {
  * axis = 2 (matrices)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorSlicingConstructorAxis2() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tens(data, 2, 3, 4);
@@ -186,7 +186,7 @@ TEST_F(TensorTest, tensorSlicingConstructorAxis2) {
  * axis = 1 (columns)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorSlicingConstructorAxis1() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -211,7 +211,7 @@ TEST_F(TensorTest, tensorSlicingConstructorAxis1) {
  * axis = 0 (columns)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorSlicingConstructorAxis0() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -235,7 +235,7 @@ TEST_F(TensorTest, tensorSlicingConstructorAxis0) {
  * Tensor: Upload data
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorUpload() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(2, 3, 4);
@@ -258,7 +258,7 @@ TEST_F(TensorTest, tensorUpload) {
  * Tensor: deviceCopyTo
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorDeviceCopyTo() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -282,7 +282,7 @@ TEST_F(TensorTest, tensorDeviceCopyTo) {
  * Tensor: Frobenius dot product
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorDotF(T epsilon) {
     // as vectors
     std::vector<T> dataA = TENSOR_DATA_234A;
@@ -307,7 +307,7 @@ TEST_F(TensorTest, tensorDotF) {
  * Tensor: Frobenius norm
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorNormF(T epsilon) {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -324,7 +324,7 @@ TEST_F(TensorTest, tensorNormF) {
  * all elements
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorSumAbs() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -341,7 +341,7 @@ TEST_F(TensorTest, tensorNormFtensorSumAbs) {
  * e.g., t(2, 3, 4)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorBracketOperator() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -360,7 +360,7 @@ TEST_F(TensorTest, tensorBracketOperator) {
  * Tensor assignment operator
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorAssignmentOperator() {
     std::vector<T> data = TENSOR_DATA_234A;
     DTensor<T> tenz(data, 2, 3, 4);
@@ -382,7 +382,7 @@ TEST_F(TensorTest, tensorAssignmentOperator) {
  * Tensor times-equals scalar
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorTimesEqualsScalar() {
     std::vector<T> data = TENSOR_DATA_234A;
     std::vector<T> dataTimes3 = {3, 6, 9, 12, 15, 18, 21, 24, 27, 24, 21, 30, 15, 12, 9, 6, 3, -3, 12, 9, 12, 9, 12,
@@ -403,7 +403,7 @@ TEST_F(TensorTest, tensorTimesEqualsScalar) {
  * Scalar times tensor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorTimesScalar() {
     std::vector<T> data = TENSOR_DATA_234A;
     std::vector<T> dataTimes3 = {3, 6, 9, 12, 15, 18, 21, 24, 27, 24, 21, 30, 15, 12, 9, 6, 3, -3, 12, 9, 12, 9, 12,
@@ -424,7 +424,7 @@ TEST_F(TensorTest, tensorTimesScalar) {
  * Tensor plus-equals tensor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorPlusEqualsTensor() {
     std::vector<T> dataA = TENSOR_DATA_234A;
     std::vector<T> dataB = TENSOR_DATA_234B;
@@ -446,7 +446,7 @@ TEST_F(TensorTest, tensorPlusEqualsTensor) {
  * Tensor minus-equals tensor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorMinusEqualsTensor() {
     std::vector<T> dataA = TENSOR_DATA_234A;
     std::vector<T> dataB = TENSOR_DATA_234B;
@@ -468,7 +468,7 @@ TEST_F(TensorTest, tensorMinusEqualsTensor) {
  * Tensor + Tensor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorPlusTensor() {
     std::vector<T> dataA = TENSOR_DATA_234A;
     std::vector<T> dataB = TENSOR_DATA_234B;
@@ -490,7 +490,7 @@ TEST_F(TensorTest, tensorPlusTensor) {
  * Tensor - Tensor
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorMinusTensor() {
     std::vector<T> dataA = TENSOR_DATA_234A;
     std::vector<T> dataB = TENSOR_DATA_234B;
@@ -512,7 +512,7 @@ TEST_F(TensorTest, tensorMinusTensor) {
  * Tensor: pointers to matrices (on device)
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorPointersToMatrices() {
     std::vector<T> dataA = TENSOR_DATA_234A;
     DTensor<T> A(dataA, 2, 3, 4);
@@ -536,7 +536,7 @@ TEST_F(TensorTest, tensorPointersToMatrices) {
  * Tensor: C = AB
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorAddAB() {
     std::vector<T> aData = {1, 2, 3, 4, 5, 6,
                             7, 8, 9, 10, 11, 12,
@@ -563,7 +563,7 @@ TEST_F(TensorTest, tensorAddAB) {
  * Tensor: getRows
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorGetRows() {
     std::vector<T> aData = {10.5, 25.0, 60.0,
                             -21.0, 720.0, -1.0,
@@ -595,7 +595,7 @@ TEST_F(TensorTest, tensorGetRows) {
  * Tensor: transpose
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorTranspose() {
     std::vector<T> aData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
     DTensor<T> A(aData, 3, 2, 2);
@@ -629,7 +629,7 @@ protected:
  * Tensor: Least squares
  * --------------------------------------- */
 
-template<typename T>
+TEMPLATE_WITH_TYPE_T
 void tensorLeastSquares1(T epsilon) {
     // TODO test with tall matrices too
     std::vector<T> aData = {1, 2,
@@ -672,8 +672,8 @@ protected:
  * and matrix rank
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void singularValuesComputation(float epsilon) {
     std::vector<T> bData = {1, 6, 6, 6, 6, 6, 6, 6,
                             2, 7, 7, 7, 7, 7, 7, 7,
@@ -699,8 +699,8 @@ TEST_F(SvdTest, singularValuesComputation) {
  * Singular values - memory mgmt
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void singularValuesMemory(float epsilon) {
     std::vector<T> bData = {1, 6, 6, 6, 6, 6, 6, 6,
                             2, 7, 7, 7, 7, 7, 7, 7,
@@ -731,8 +731,8 @@ TEST_F(SvdTest, singularValuesMemory) {
 /* ---------------------------------------
  * SVD with multiple matrices
  * --------------------------------------- */
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void singularValuesMultipleMatrices(float epsilon) {
     std::vector<T> aData = {1, 2, 3, 4, 5, 6, 1, 1, 1, 2, 2, 2, 0, 0, 0, 0, 0, 1};
     DTensor<T> A(aData, 3, 2, 3);
@@ -779,8 +779,8 @@ TEST_F(SvdTest, singularValuesMultipleMatrices) {
  * SVD for rank computation of multiple
  * matrices
  * --------------------------------------- */
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void singularValuesRankMultipleMatrices(float epsilon) {
     std::vector<T> aData = {1, 4, 7, 10, 2, 5, 8, 11, 3, 6, 9, 0,
                             1, 4, 7, 10, 2, 5, 8, 11, 3, 6, 9, 12,
@@ -815,8 +815,8 @@ protected:
  * Cholesky factorisation
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void choleskyFactorisation(T epsilon) {
     std::vector<T> aData = {10.0, 2.0, 3.0,
                             2.0, 20.0, -1.0,
@@ -838,8 +838,8 @@ TEST_F(CholeskyTest, choleskyFactorisation) {
  * Cholesky factorisation: solve system
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void choleskyFactorisationSolution(T epsilon) {
     std::vector<T> aData = {10.0, 2.0, 3.0,
                             2.0, 20.0, -1.0,
@@ -874,8 +874,8 @@ TEST_F(CholeskyTest, choleskyFactorisationSolution) {
  * Batched Cholesky factorisation
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void choleskyBatchFactorisation(T epsilon) {
     std::vector<T> aData = {10.0, 2.0, 3.0,
                             2.0, 20.0, -1.0,
@@ -906,8 +906,8 @@ TEST_F(CholeskyTest, choleskyBatchFactorisation) {
  * Batched Cholesky solve
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void choleskyBatchFactorSolve(T epsilon) {
     std::vector<T> aData = {10.0, 2.0, 3.0,
                             2.0, 20.0, -1.0,
@@ -947,8 +947,8 @@ TEST_F(CholeskyTest, choleskyBatchFactorSolve) {
  * Batched Cholesky solve (factor provided)
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void choleskyBatchSolve(T epsilon) {
     std::vector<T> aData = {10.0, 2.0, 3.0,
                             2.0, 20.0, -1.0,
@@ -1007,8 +1007,8 @@ protected:
  * Basic nullspace test
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void computeNullspaceTensor(T epsilon) {
     std::vector<T> aData = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 0,
                             1, 2, 3, 4, 5, 6, 7, 8, 9, 7, 8, 9,
@@ -1048,8 +1048,8 @@ TEST_F(NullspaceTest, computeNullspaceTensor) {
  * Nullspace is trivial
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void computeNullspaceTrivial(T epsilon) {
     std::vector<T> data{4, 5, 7,
                         4, 1, 8,
@@ -1072,8 +1072,8 @@ TEST_F(NullspaceTest, computeNullspaceTrivial) {
  * Project onto nullspace
  * --------------------------------------- */
 
-template<typename T>
-requires std::floating_point<T>
+TEMPLATE_WITH_TYPE_T
+TEMPLATE_CONSTRAINT_REQUIRES_FPX
 void projectOnNullspaceTensor(T epsilon) {
     // offline
     size_t m = 3;

